<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Microbiome | Jinwoo Cho </title> <meta name="author" content="Jinwoo Cho"> <meta name="description" content="Introductory project from SNU"> <meta name="keywords" content="Jinwoo_Cho, portfolio-website"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%F0%9F%8E%93&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://kevinjwcho.github.io/projects/appli_2/"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script src="/assets/js/theme.js?0afe9f0ae161375728f7bcc5eb5b4ab4"></script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Jinwoo</span> Cho </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">About </a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">CV </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">Publications </a> </li> <li class="nav-item active"> <a class="nav-link" href="/projects/">Projects <span class="sr-only">(current)</span> </a> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="fa-solid fa-moon"></i> <i class="fa-solid fa-sun"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="post"> <header class="post-header"> <h1 class="post-title">Microbiome</h1> <p class="post-description">Introductory project from SNU</p> </header> <article> <h2 id="microbiome-data-analysis-tutorial">Microbiome Data Analysis Tutorial</h2> <p>This is a microbiome data analysis tutorial for the lab that I made.</p> <h3 id="library">Library</h3> <p>The necessary libraries are following.</p> <pre><code class="language-{r}"># BiocManager::install("phyloseq")
library(phyloseq)
library(dplyr)
library(data.table)
library(stringr)
</code></pre> <h3 id="load-biome-data">Load .biome data</h3> <p>There are four forms of phyloseq class data.</p> <ol> <li>otu_table : OTU count data</li> <li>sample_data : Sample data</li> <li>tax_table : Taxonomy data</li> <li>phy_tree : Phylogentic Tree data</li> </ol> <h3 id="generate-phylogenetic-tree">Generate Phylogenetic Tree</h3> <p>Using the information of Denovo, we can generate the phylogenetic tree.</p> <ol> <li>adegenet : A package that includes the function, <em>fasta2DNAbin</em>, loading <code class="language-plaintext highlighter-rouge">.fasta</code> data.</li> <li>ape : This package can generate <code class="language-plaintext highlighter-rouge">Phylogentic Tree</code> </li> </ol> <pre><code class="language-{r}"># install.packages("ape")
# install.packages("adegenet")
library(adegenet)
library(ape)
</code></pre> <p>Then you can load <em>fasta</em> data.</p> <pre><code class="language-{r}">dna &lt;-fasta2DNAbin("F:/work/Microbiome_EDC/HN00117382_Reports/otus_rep.fasta")
class(dna)
dna
</code></pre> <p>There are several options to create Tree. You can check from <em>?dist.dna()</em>. Importantly, if there is missing value in the distance matrix, you can only use <em>njs()</em> function. Otherwise, <em>nj()</em> can be applicable.</p> <p><strong>It takes longer than your thonght, so please save the results.</strong></p> <pre><code class="language-{r}">D &lt;- dist.dna(dna, model = "raw")
length(D)
tre &lt;- nj(D)

D &lt;- dist.dna(dna, model = "TN93")
length(D)
tre &lt;- njs(D)

save(tre, file = "F:/work/Microbiome_EDC/RData/TN93_tre.RData")
</code></pre> <p>You can load your results from the following code.</p> <pre><code class="language-{r}">load("F:/work/Microbiome_EDC/RData/TN93_tre.RData")
</code></pre> <p>You can assign the loaded tree to the phyloseq form data.</p> <pre><code class="language-{r}">tre$tip.label &lt;- sapply(strsplit(tre$tip.label, "\t"), function(x)x[1])
phy_tree(data) &lt;- tre
data
</code></pre> <p>Lastly, if you want to separate your phyloseq data, do the following.</p> <pre><code class="language-{r}">otu &lt;- otu_table(data)
sample_dat &lt;- sample_data(data)
tax &lt;- tax_table(data)
phy_tre &lt;- phy_tree(data)
</code></pre> <h3 id="managing-phyloseq-data">Managing phyloseq data</h3> <p>You can check the ranking level of taxonomy by <em>rank_names()</em> function. ```{r, results = ‘markup’} rank_names(data) tax_table(data) %&gt;% colnames</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>And add the label of the ranking.
```{r, results='markup'}
colnames(tax_table(data))&lt;- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
rank_names(data)
</code></pre></div></div> <h3 id="bar-graph">Bar graph</h3> <p>First of all, you need <em>otu_table</em> to draw a bar graph to see the abundance of each sample. Besides, since the total count of microbiome for each sample is different, if you want to proportion of OTU through a bar graph, you need to compute the relative frequency of the count. To convert the count value, you can do following.</p> <pre><code class="language-{r}">rel_sum &lt;- function(x) round((x/sum(x)))
rel_data = transform_sample_counts(data, rel_sum)
total = median(sample_sums(data))
standf &lt;- function(x, t=total) round(t*(x/sum(x)))
stand_data &lt;- transform_sample_counts(data, standf)
otu_table(data)[1:10, 1:5]
otu_table(rel_data)[1:10, 1:5]
</code></pre> <p>You can draw a bar graph using <em>ggplot2</em> package.</p> <pre><code class="language-{r}">library(ggplot2)
plot_bar(stand_data, fill = "Phylum")+
  geom_bar(aes(color=Phylum, fill=Phylum), stat="identity", position = "stack")+
  theme(axis.text.x = element_blank())
</code></pre> <div class="row"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/img2_2-480.webp 480w,/assets/img/img2_2-800.webp 800w,/assets/img/img2_2-1400.webp 1400w," sizes="95vw" type="image/webp"></source> <img src="/assets/img/img2_2.jpg" class="img-fluid rounded z-depth-1" width="100%" height="auto" title="example image" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <h3 id="heatmap">Heatmap</h3> <p>If your data has the larger number of taxonomy, it would be better to filter your data in heatmap analysis. The following code is to extract the taxonomy whish has the larger counts than median by 10\%.</p> <pre><code class="language-{r}">stand_data_abund &lt;- filter_taxa(stand_data, function(x) sum(x&gt;total*0.1)&gt;0, TRUE)
stand_data_abund
</code></pre> <p>When you create a heatmap, the result is following. ```{r, echo = F} plot_heatmap(stand_data_abund, method = “NMDS”, distance = “bray”)+theme(axis.text.x = element_blank())</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;div class="row"&gt;
    &lt;div class="col-sm mt-3 mt-md-0"&gt;
        

&lt;figure
  
&gt;
  &lt;picture&gt;
    &lt;!-- Auto scaling with imagemagick --&gt;
    &lt;!--
      See https://www.debugbear.com/blog/responsive-images#w-descriptors-and-the-sizes-attribute and
      https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images for info on defining 'sizes' for responsive images
    --&gt;
    
      &lt;source
        class="responsive-img-srcset"
        srcset="/assets/img/img2_3-480.webp 480w,/assets/img/img2_3-800.webp 800w,/assets/img/img2_3-1400.webp 1400w,"
        
          sizes="95vw"
        
        type="image/webp"
      &gt;
    
    &lt;img
      src="/assets/img/img2_3.jpg"
      
        class="img-fluid rounded z-depth-1"
      
      
        width="100%"
      
      
        height="auto"
      
      
      
      
        title="example image"
      
      
      
        loading="eager"
      
      onerror="this.onerror=null; $('.responsive-img-srcset').remove();"
    &gt;
  &lt;/picture&gt;

  
&lt;/figure&gt;

    &lt;/div&gt;
&lt;/div&gt;


### Alpha diversity

To compute alpha diversity, you can use *estimate_richness()* to obtain Chao1, Shannon, and Simpson.
```{r, results = 'markup'}
estimate_richness(data, measures = "Chao1") %&gt;% head
estimate_richness(data, measures = "Shannon") %&gt;% head
estimate_richness(data, measures = "Simpson") %&gt;% head
</code></pre></div></div> <pre><code class="language-{r}">plot_richness(data, measures = c("Chao1", "Shannon"))+theme(axis.text.x = element_blank())
</code></pre> <div class="row"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/img2_4-480.webp 480w,/assets/img/img2_4-800.webp 800w,/assets/img/img2_4-1400.webp 1400w," sizes="95vw" type="image/webp"></source> <img src="/assets/img/img2_4.jpg" class="img-fluid rounded z-depth-1" width="100%" height="auto" title="example image" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <h3 id="beta-diversity">Beta diversity</h3> <p>You can compute beta diversity by <em>distance()</em> function. ```{r, results = ‘markup’} distance(data, method = “unifrac”) %&gt;% head distance(data, method = “wunifrac”) %&gt;% head distance(data, method = “bray”) %&gt;% head distance(data, method = “dpcoa”) %&gt;% head</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
Since the beta diversity is a distance matrix, we need to reduce dimension using ordination. Ordinatiom includes methods *NMDS* and *PCoA*. 
```{r, results = 'markup'}
data.ord1 &lt;- ordinate(data, "NMDS", "bray")
data.ord2 &lt;- ordinate(data, "NMDS", "wunifrac")
data.ord3 &lt;- ordinate(data, "PCoA", "bray")
data.ord4 &lt;- ordinate(data, "PCoA", "wunifrac")
</code></pre></div></div> <pre><code class="language-{r}">plot_ordination(data, data.ord1, type="samples", title = "Samples", color = "baby_gen")+
  geom_point(size=2)
</code></pre> <div class="row"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/img2_5-480.webp 480w,/assets/img/img2_5-800.webp 800w,/assets/img/img2_5-1400.webp 1400w," sizes="95vw" type="image/webp"></source> <img src="/assets/img/img2_5.jpg" class="img-fluid rounded z-depth-1" width="100%" height="auto" title="example image" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> </article> </div> </div> <footer class="fixed-bottom" role="contentinfo"> <div class="container mt-0"> © Copyright 2025 Jinwoo Cho. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2930004b8d7fcd0a8e00fdcfc8fc9f24"></script> <script defer src="/assets/js/common.js?da39b660470d1ba6e6b8bf5f37070b6e"></script> <script defer src="/assets/js/copy_code.js?12775fdf7f95e901d7119054556e495f" type="text/javascript"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script type="text/javascript">window.MathJax={tex:{tags:"ams"}};</script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.js"></script> <script defer src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script type="text/javascript">function progressBarSetup(){"max"in document.createElement("progress")?(initializeProgressElement(),$(document).on("scroll",function(){progressBar.attr({value:getCurrentScrollPosition()})}),$(window).on("resize",initializeProgressElement)):(resizeProgressBar(),$(document).on("scroll",resizeProgressBar),$(window).on("resize",resizeProgressBar))}function getCurrentScrollPosition(){return $(window).scrollTop()}function initializeProgressElement(){let e=$("#navbar").outerHeight(!0);$("body").css({"padding-top":e}),$("progress-container").css({"padding-top":e}),progressBar.css({top:e}),progressBar.attr({max:getDistanceToScroll(),value:getCurrentScrollPosition()})}function getDistanceToScroll(){return $(document).height()-$(window).height()}function resizeProgressBar(){progressBar.css({width:getWidthPercentage()+"%"})}function getWidthPercentage(){return getCurrentScrollPosition()/getDistanceToScroll()*100}const progressBar=$("#progress");window.onload=function(){setTimeout(progressBarSetup,50)};</script> </body> </html>